### PR TITLE
[RFC] Refactor: remove uses of has_mbyte in option.c

### PR DESCRIPTION
**Pull request by kranki**
- Add test coverage to functions using has_mbyte
- Switch to using MB_PTR2LEN et al.

### CODE REVIEW

**Comment by bfredl** [comment]
> MB_PTR2LEN

why not `mb_ptr2len()` ?

**Comment by kranki** [comment]
MB_PTR2LEN lets us eliminate the else branch.

**Comment by bfredl** [comment]
`mb_ptr2len()` (and `mb_ptr2char()`) is always correct. `MB_PTR2LEN` was an optimization specifically for `enc=latin1` which is deprecated, so  right now it is a deoptimization... And in #2905 MB_PTR2LEN is right now defined to `mb_ptr2len()` (but eventually it should just be changed to `mb_ptr2len()` at use)

**Comment by kranki** [comment]
I get it. I'm trying to help with #2905. My reasoning goes like this:
Using MB_PTR2LEN can be used to merge multibyte and single byte char handling. 
My latest [commit]( https://github.com/neovim/neovim/pull/5298/commits/c0fb53db0f68e0c40842e7660562749eef055f57) is an example of which the codebase is full of.
After #2905 is ready MB_PTR2LEN can be simply changed to mb_ptr2len with a simple rename.

**Comment by kranki** [comment]
It is a deoptimization but I don't think it unreasonable. It is not a place for optimization. Setting 'matchpair' can take the couple of extra if(has_mbyte) if the compiler doesn't work that out.

**Comment by kranki** [comment]
> why not mb_ptr2len() ?

Or maybe I misunderstood? Does `mb_ptr2len` return 1 when `has_mbyte == true` ?

**Comment by bfredl** [comment]
> Or maybe I misunderstood? Does mb_ptr2len return 1 when has_mbyte == true ?

You meant `== false?` You can eliminate the `if(has_mbyte)` and still use `mb_ptr2len()`, switching to `MB_PTR2LEN` does not gain anything.

**Comment by kranki** [comment]
Yes `==false,` I edited the comment right after but I guess you got the unedited.
I will use `mb_ptr2len` then.

Do you know some lua test to serve as an example of passing multibyte characters in? I'm doing these refactorings to learn to write tests, so that would be helpful.

**Comment by bfredl** [comment]
The important distinction was really that`mb_ptr2len` was only available when vim was compiled with `+mbyte` (and all those `if (has_mbyte)` was also guarded with `#ifdef FEAT_MBYTE`) but `MB_PTR2LEN` was always available. As you see lot's of cleanups are already possible just because `FEAT_MBYTE` is now hardcoded, but feel free to assume #2905 if that makes stuff easier, we should get that one merged soon anyway.

> Do you know some lua test to serve as an example of passing multibyte characters in? 

In general you should be able to have utf-8 chars verbatim in strings in lua code. see `ui/bufhl_spec.lua` (the last test) and `api/vim_spec.lua` (strwidth) for some examples. (Adding tests for `enc=utf8`, which tests use by default if not changed, is enough, as it is the only mode we will support moving forward.)

File: src/nvim/option.c [hunk 1]
```diff
@@ -3211,6 +3191,33 @@ did_set_string_option (
   return errmsg;
 }
 
+static bool mps_valid( char_u *mps )
+{
+  for (char_u *p = mps; *p != NUL; p++) {
+    if (*p != NUL) {
```
Comment (oni-link) [inline src/nvim/option.c]: This check could be removed, we already know that `*p` is not `NUL`.

File: src/nvim/option.c [hunk 2]
```diff
@@ -3211,6 +3191,33 @@ did_set_string_option (
...
+
+    if (*p == NUL) {
+      return false;
+    }
+    p += mb_ptr2len(p);
+
+    if (*p != NUL && *p != ',') {
+      return false;
+    }
+    if (*p == NUL) {
```
Comment (oni-link) [inline src/nvim/option.c]: This block could be removed.

File: src/nvim/option.c [hunk 3]
```diff
@@ -1607,16 +1607,17 @@ do_set (
...
+                  arg++;   // remove backslash
+                }
+                if ((i = mb_ptr2len(arg)) > 1) {
+                  // copy multibyte char
                   memmove(s, arg, (size_t)i);
                   arg += i;
                   s += i;
-                } else
+                } else {
                   *s++ = *arg++;
```
Comment (bfredl) [inline src/nvim/option.c]: I think this branch can be removed. The optimizer can already reason `memmove`/`memcpy`, and it's not like this code is super hot anyway...

File: src/nvim/option.c [hunk 4]
```diff
@@ -3221,6 +3184,31 @@ did_set_string_option (
...
+
+    if (*p == NUL) {
+      return false;
+    }
+    p += mb_ptr2len(p);
+
+    if (*p != NUL && *p != ',') {
+      return false;
+    }
+  }
```
Comment (oni-link) [inline src/nvim/option.c]: Did I comment here to remove the `NUL` check? The check cannot be removed if `p++` is used in the loop header. For `"(:)"` the loop would check for `NUL` one byte after the C string.

**Comment by janlazo** [comment]
`has_mbyte` is removed from option.c in https://github.com/neovim/neovim/pull/13275

**Pull request closed without merging**